packages/ui/components/form/selectimproved/components/Options.tsx

```diff
@@ -0,0 +1,126 @@
+import React, { useCallback, useContext, useEffect, useMemo, useRef, useState } from "react";
+
+import { useKeyPress } from "@calcom/lib/hooks/useKeyPress";
+
+import { Label } from "../../inputs/Label";
+import Item from "./Item";
+import { SelectContext } from "./SelectProvider";
+import { Option } from "./type";
+
+interface OptionsProps<T extends Option> {
+  list: T[];
+  inputValue: string;
+  isMultiple: boolean;
+  selected: T | T[] | null;
+  searchBoxRef: React.RefObject<HTMLInputElement>;
+}
+
+type FlattenedOption = Option & { current: number; groupedIndex?: number };
+
+const flattenOptions = (options: Option[], groupCount?: number): FlattenedOption[] => {
+  return options.reduce((acc, option, current) => {
+    if (option.options) {
+      return [...acc, ...flattenOptions(option.options, current + (groupCount || 0))];
+    }
+    return [...acc, { ...option, current, groupedIndex: groupCount }];
+  }, [] as FlattenedOption[]);
+};
+
+function FilteredItem<T extends Option>({
+  index,
+  keyboardFocus,
+  item,
+  inputValue,
+  list,
+}: {
+  index: number;
+  keyboardFocus: number;
+  item: FlattenedOption;
+  inputValue: string;
+  list: T[];
+}) {
+  const focused = index === keyboardFocus;
+  const ref = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    if (ref.current && focused) {
+      ref.current.scrollIntoView({ behavior: "smooth" });
+    }
+  }, [ref, focused]);
+
+  return (
+    <div className="px-2.5" key={index} ref={ref}>
+      {item.current === 0 && item.groupedIndex !== undefined && !inputValue && (
+        <Label>{list[item.groupedIndex].label}</Label>
+      )}
+      <Item item={item} index={index} focused={focused} />
+    </div>
+  );
+}
+
+function Options<T extends Option>({ list, inputValue, searchBoxRef }: OptionsProps<T>) {
+  const { classNames, handleValueChange } = useContext(SelectContext);
+  const [keyboardFocus, setKeyboardFocus] = useState(-1);
+  const enterPress = useKeyPress("Enter", searchBoxRef);
+
+  const flattenedList = useMemo(() => flattenOptions(list), [list]);
+
+  const totalOptionsLength = useMemo(() => {
+    return flattenedList.length;
+  }, [flattenedList]);
+
+  useKeyPress("ArrowDown", searchBoxRef, () => setKeyboardFocus((prev) => (prev + 1) % totalOptionsLength));
+  useKeyPress("ArrowUp", searchBoxRef, () =>
+    setKeyboardFocus((prev) => (prev - 1 + totalOptionsLength) % totalOptionsLength)
+  );
+
+  useEffect(() => {
+    if (enterPress) {
+      const item = filteredList[keyboardFocus];
+      if (!item || item.disabled) return;
+      handleValueChange(item);
+    }
+    // We don't want to re-run this effect when handleValueChange changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [enterPress, keyboardFocus, list]);
```

Comment:
Had to leave this as the old hook as we dont want this to run when handleValueChanges